packages/frontend/core/src/modules/workbench/workbench.ts

```diff
@@ -0,0 +1,91 @@
+import { Unreachable } from '@affine/env/constant';
+import { LiveData } from '@toeverything/infra';
+import type { To } from 'history';
+import { combineLatest, map, switchMap } from 'rxjs';
+
+import { View } from './view';
+
+export type WorkbenchPosition = 'beside' | 'active' | number;
+
+export class Workbench {
+  readonly views = new LiveData([new View()]);
+
+  activeViewIndex = new LiveData(0);
+  activeView = LiveData.from(
+    combineLatest([this.views, this.activeViewIndex]).pipe(
+      map(([views, index]) => views[index])
+    ),
+    this.views.value[this.activeViewIndex.value]
+  );
+
+  location = LiveData.from(
+    this.activeView.pipe(switchMap(view => view.location)),
+    this.views.value[this.activeViewIndex.value].history.location
+  );
+
+  active(index: number) {
+    this.activeViewIndex.next(index);
+  }
+
+  createView(at: WorkbenchPosition = 'beside') {
+    const view = new View();
+    const newViews = [...this.views.value];
+    newViews.splice(this.indexAt(at), 0, view);
+    this.views.next(newViews);
+    return newViews.indexOf(view);
+  }
+
+  open(to: To, { at = 'active' }: { at?: WorkbenchPosition } = {}) {
+    let view = this.viewAt(at);
+    if (!view) {
+      const newIndex = this.createView(at);
+      view = this.viewAt(newIndex);
+      if (!view) {
+        throw new Unreachable();
+      }
+    }
+    view.history.push(to);
```

Comment:
replace will not be supported?

Reply:
In some cases, a route view may redirect to another view on landing, which creates a unwanted item in the history stack.
e.g., `goto("root.a") =>  "root.a" redirected to "root.a.foo"`
By using `replace`, "root.a" will not be on the stack. If not using `replace` and using route backwards at `root.a.foo`, the app will bounce back from `root.a` because it triggers redirect again

Reply:
replace is now supported !